internal/storage/sql/common/rollout.go

```diff
@@ -323,79 +327,111 @@ func (s *Store) CreateRollout(ctx context.Context, r *flipt.CreateRolloutRequest
 			CreatedAt:    now,
 			UpdatedAt:    now,
 		}
 	)
 
+	switch r.GetRule().(type) {
+	case *flipt.CreateRolloutRequest_Segment:
+		rollout.Type = flipt.RolloutType_SEGMENT_ROLLOUT_TYPE
+	case *flipt.CreateRolloutRequest_Percentage:
+		rollout.Type = flipt.RolloutType_PERCENTAGE_ROLLOUT_TYPE
+	case nil:
+		return nil, errs.ErrInvalid("rollout rule is missing")
+	default:
+		return nil, errs.ErrInvalidf("invalid rollout rule type %T", r.GetRule())
+	}
+
 	tx, err := s.db.Begin()
 	if err != nil {
 		return nil, err
 	}
 
+	defer func() {
+		if err != nil {
+			_ = tx.Rollback()
+		}
+	}()
+
 	if _, err := s.builder.Insert(tableRollouts).
 		RunWith(tx).
 		Columns("id", "namespace_key", "flag_key", "\"type\"", "rank", "description", "created_at", "updated_at").
 		Values(rollout.Id, rollout.NamespaceKey, rollout.FlagKey, rollout.Type, rollout.Rank, rollout.Description,
 			&fliptsql.Timestamp{Timestamp: rollout.CreatedAt},
 			&fliptsql.Timestamp{Timestamp: rollout.UpdatedAt},
 		).ExecContext(ctx); err != nil {
-		_ = tx.Rollback()
 		return nil, err
 	}
 
-	switch r.GetType() {
-	case flipt.RolloutType_SEGMENT_ROLLOUT_TYPE:
+	switch r.GetRule().(type) {
+	case *flipt.CreateRolloutRequest_Segment:
+		rollout.Type = flipt.RolloutType_SEGMENT_ROLLOUT_TYPE
+
 		var segmentRule = r.GetSegment()
 
 		if _, err := s.builder.Insert(tableRolloutSegments).
 			RunWith(tx).
 			Columns("id", "rollout_id", "namespace_key", "segment_key", "\"value\"").
 			Values(uuid.Must(uuid.NewV4()).String(), rollout.Id, rollout.NamespaceKey, segmentRule.SegmentKey, segmentRule.Value).
 			ExecContext(ctx); err != nil {
-			_ = tx.Rollback()
 			return nil, err
 		}
 
 		rollout.Rule = &flipt.Rollout_Segment{
 			Segment: segmentRule,
 		}
-	case flipt.RolloutType_PERCENTAGE_ROLLOUT_TYPE:
+	case *flipt.CreateRolloutRequest_Percentage:
+		rollout.Type = flipt.RolloutType_PERCENTAGE_ROLLOUT_TYPE
+
 		var percentageRule = r.GetPercentage()
 
 		if _, err := s.builder.Insert(tableRolloutPercentages).
 			RunWith(tx).
 			Columns("id", "rollout_id", "namespace_key", "percentage", "\"value\"").
 			Values(uuid.Must(uuid.NewV4()).String(), rollout.Id, rollout.NamespaceKey, percentageRule.Percentage, percentageRule.Value).
 			ExecContext(ctx); err != nil {
-			_ = tx.Rollback()
 			return nil, err
 		}
 
 		rollout.Rule = &flipt.Rollout_Percentage{
 			Percentage: percentageRule,
 		}
 	default:
-		_ = tx.Rollback()
-		return nil, fmt.Errorf("invalid rollout rule type %v", r.GetType())
+		return nil, fmt.Errorf("invalid rollout rule type %v", rollout.Type)
 	}
 
 	return rollout, tx.Commit()
 }
 
-func (s *Store) UpdateRollout(ctx context.Context, r *flipt.UpdateRolloutRequest) (*flipt.Rollout, error) {
+func (s *Store) UpdateRollout(ctx context.Context, r *flipt.UpdateRolloutRequest) (_ *flipt.Rollout, err error) {
 	if r.NamespaceKey == "" {
 		r.NamespaceKey = storage.DefaultNamespace
 	}
 
+	if r.Id == "" {
+		return nil, errs.ErrInvalid("rollout ID not supplied")
```

Comment:
we usually do validation in the `validation.go` file for `XRequest` types so at the server layer, is there reason we are doing it here in the storage layer as well?

Reply:
That is fair, I guess not really. This came about because when I introduced the fetch to ensure the type on the rollout, there was an existing test which was not adding an Id, but expecting a different validation error to occur. However, that validation happened now after we try to get the rollout by Id, so I got a hard to reason about error in that test case: `rollout "default/" not found`. Which I thought was a bit obscure, so I added this for clarity.
I can definitely drop it and assume not nil.

Reply:
im fine w leaving it, just wanted to check. prob could just have validation in the storage layer in rare cases like this but mostly defer to the server validation logic